internal/controllers/machine_controller.go

```diff
@@ -98,24 +101,30 @@ func NewMachineReconciler(
 	}
 
 	return &MachineReconciler{
-		log:                            log,
-		queue:                          workqueue.NewRateLimitingQueue(workqueue.DefaultControllerRateLimiter()),
-		libvirt:                        libvirt,
-		machines:                       machines,
-		machineEvents:                  machineEvents,
-		EventRecorder:                  eventRecorder,
-		guestCapabilities:              opts.GuestCapabilities,
-		tcMallocLibPath:                opts.TCMallocLibPath,
-		host:                           opts.Host,
-		imageCache:                     opts.ImageCache,
-		raw:                            opts.Raw,
-		volumePluginManager:            opts.VolumePluginManager,
-		networkInterfacePlugin:         opts.NetworkInterfacePlugin,
-		resyncIntervalVolumeSize:       opts.ResyncIntervalVolumeSize,
-		resyncIntervalGarbageCollector: opts.ResyncIntervalGarbageCollector,
-		enableHugepages:                opts.EnableHugepages,
-		gcVMGracefulShutdownTimeout:    opts.GCVMGracefulShutdownTimeout,
-		volumeCachePolicy:              opts.VolumeCachePolicy,
+		log: log,
+		queue: workqueue.NewRateLimitingQueueWithConfig(workqueue.DefaultControllerRateLimiter(), workqueue.RateLimitingQueueConfig{
+			MetricsProvider: metrics.WorkqueueMetricsProvider{},
+			Name:            MachineReconcilerName,
+		}),
+		libvirt:                                 libvirt,
+		machines:                                machines,
+		machineEvents:                           machineEvents,
+		EventRecorder:                           eventRecorder,
+		guestCapabilities:                       opts.GuestCapabilities,
+		tcMallocLibPath:                         opts.TCMallocLibPath,
+		host:                                    opts.Host,
+		imageCache:                              opts.ImageCache,
+		raw:                                     opts.Raw,
+		volumePluginManager:                     opts.VolumePluginManager,
+		networkInterfacePlugin:                  opts.NetworkInterfacePlugin,
+		resyncIntervalVolumeSize:                opts.ResyncIntervalVolumeSize,
+		resyncIntervalGarbageCollector:          opts.ResyncIntervalGarbageCollector,
+		enableHugepages:                         opts.EnableHugepages,
+		gcVMGracefulShutdownTimeout:             opts.GCVMGracefulShutdownTimeout,
+		volumeCachePolicy:                       opts.VolumeCachePolicy,
+		metricsReconcileDuration:                metrics.ControllerRuntimeReconcileDuration.WithLabelValues(MachineReconcilerName),
+		metricsControllerRuntimeActiveWorker:    metrics.ControllerRuntimeActiveWorker.WithLabelValues(MachineReconcilerName),
+		metricsControllerRuntimeReconcileErrors: metrics.ControllerRuntimeReconcileErrors.WithLabelValues(MachineReconcilerName),
```

Comment:
Why are the metrics not passed as options? It makes it hard to pass a mock metric provider in tests

Reply:
why do you want to mock metrics? these metrics has to represent internal stuff of controller (errors, workers, ...) and they don't have any special logic so you can create them in unit tests and it will work.

Reply:
You can pass your own simple implementation in order to check the functionality.

Reply:
i checked that and i don't think we need mocking interface.

You can very nice test metrics directly with testutil package from prometheus without web server:

```
var _ = Describe("Controller runtime metrics", func() {
	Context("metric ControllerRuntimeReconcileErrors", func() {
		It("should increment metric", func() {
			m := metrics.ControllerRuntimeReconcileErrors.WithLabelValues(controllers.MachineReconcilerName)
			m.Inc()
			m.Inc()
			expectedMetric := `
# HELP controller_runtime_reconcile_errors_total Total number of reconciliation errors per controller
# TYPE controller_runtime_reconcile_errors_total counter
controller_runtime_reconcile_errors_total{controller="machine-reconciler"} 2
`
			reader := strings.NewReader(expectedMetric)
			Expect(testutil.GatherAndCompare(prometheus.DefaultGatherer, reader, "controller_runtime_reconcile_errors_total")).ToNot(HaveOccurred())
		})
	})
})
```

Reply:
The code above describes a test case which checks if you can increment a metrics counter, which is covered already by the lib. 
I'm talking about a test where we create a resource, and check if metric `x`, or `y` is as expected.

Reply:
you have it described above.

Reply:
applied in https://github.com/ironcore-dev/libvirt-provider/pull/384/commits/faaa29bd13c03b74f24feeedd931a2b98430a23b